src/tools/compiletest/src/runtest.rs

```diff
@@ -2406,16 +2406,60 @@ impl<'test> TestCx<'test> {
             .args(&self.props.cbmc_flags);
         self.add_rmc_dir_to_path(&mut rmc);
         let proc_res = self.compose_and_run_compiler(rmc, None);
-        // Print an error if the verification result is not expected.
-        if self.should_compile_successfully(self.pass_mode()) {
-            if !proc_res.status.success() {
-                self.fatal_proc_rec("test failed: expected success, got failure", &proc_res);
+        // If the test file contains expected failures in some locations, ensure
+        // that verification does indeed fail in those locations
+        let fail_lines = TestCx::find_expect_fail_lines(&self.testpaths.file);
+        if !fail_lines.is_empty() {
+            if let Some(num) = TestCx::verify_expect_fail_in_lines(&proc_res.stdout, fail_lines) {
+                self.fatal_proc_rec(
+                    &format!("test failed: expected failure in line {}, got success", num),
+                    &proc_res,
+                )
             }
         } else {
-            if proc_res.status.success() {
-                self.fatal_proc_rec("test failed: expected failure, got success", &proc_res);
+            // Print an error if the verification result is not expected.
+            if self.should_compile_successfully(self.pass_mode()) {
+                if !proc_res.status.success() {
+                    self.fatal_proc_rec("test failed: expected success, got failure", &proc_res);
+                }
+            } else {
+                if proc_res.status.success() {
+                    self.fatal_proc_rec("test failed: expected failure, got success", &proc_res);
+                }
+            }
+        }
+    }
+
+    /// Verify that `str` contains expected failures corresponding to each line in `lines`.
+    fn verify_expect_fail_in_lines(str: &str, line_nums: Vec<usize>) -> Option<usize> {
+        let mut lines = str.lines();
+        let first_fail = &format!("line {} EXPECTED FAIL", line_nums[0]);
+        let mut line = lines.find(|&line| line.contains(first_fail));
+
+        for line_num in line_nums {
+            // It is safe to call unwrap here because RMC always prints an
+            // "EXPECTED FAIL" line for each call to `__VERIFIER_expect_fail`.
+            if line.unwrap().contains("SUCCESS") {
+                return Some(line_num);
+            }
+            // "EXPECTED FAIL" lines are printed in sequence, so we can
+            // immediately check the next line.
+            line = lines.next()
```

Comment:
Worried about this. Say you have expected-fail assertions in two functions from different files - they will not appear next to each other, right?

Reply:
Yes. You're right. I did not consider that. What should be the plan here? Should I look for expected failures in other files too? That complicates things quite a bit... I could also check for the file name first...

Reply:
Why not use the the `lines.find` call that you use for finding the first element?

Reply:
It does not solve this problem. You can have two `__VERIFIER_expect_fail` on different files, but the same line.

Reply:
One option is to add more lines to be checked from other files after getting the output. But you can open an issue and deal with that later. Adding a check for the file name first should be fine for this PR.

Reply:
There's another way to go about this: You can reverse the procedure to inspect files looking for expected-fail assertions within the files based on the output. I would check anyway that the main file contains expected-fail assertions (in case the text changes in the future), but then the problem is reduced to building the files-lines relation.

Reply:
Done! I added a naïve check to ensure that at least few tests fail if the output changes in the future. I opened an issue to come up with a better check #286. Building a file-line relation complicates the check. I did not implement it because it can be manually extracted from RMC's output and there is no immediate use. I opened an issue to track this as well #285.